Feature/add method to obtain radio information

mariomaz opened the pull request:
Add method to obtain radio information by issuing command
NL80211_CMD_GET_WIPHY.

This method will later be used to compute MediaType field required by
several TLVs and to fill in the HT/VHT capabilities TLVs.

mariomaz committed:
bwl: add method to obtain radio information

Add method to obtain radio information by issuing command
NL80211_CMD_GET_WIPHY.

This method will later be used to compute MediaType field required by
several TLVs and to fill in the HT/VHT capabilities TLVs.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

arnout commented on common/beerocks/bwl/include/bwl/nl80211_client.h:
```diff
@@ -33,6 +51,116 @@ namespace bwl {
+     * Information obtained with NL80211_CMD_GET_WIPHY command through a NL80211 socket.
+     * See NL80211_FREQUENCY_ATTR_* in <linux/nl80211.h> for a description of each field.
+     */
+    struct sChannelInfo {
```
Is it really necessary/useful to copy the sWifiChannel structure?

tomereli replied:
I don't see why, I think we should use the same structure here.
Same for the sBandInfo I guess.

mariomaz replied:
Replaced sChannelInfo with sWifiChannel

mariomaz replied:
According to what has been agreed on meeting, I've recovered and used sChannelInfo. 
Now there are 3 different structures containing channel information: in nl80211_client.h, in beerocks_message_structs.h and in base_wlan_hal_types.h. This is something to be fixed in a future PR

arnout commented on common/beerocks/bwl/include/bwl/nl80211_client.h:
```diff
@@ -33,6 +51,116 @@ namespace bwl {
+
+        /**
+         * Maximum channel bandwidth in MHz.
+         * Possible values are 0 (unknown), 20, 40, 80 and 160 MHz.
```
Do you know how nl80211 represents 80+80?

mariomaz replied:
No idea about WiFi. 
From code borrowed from `iw`:
```c
                        switch ((band.vht_capability >> 2) & 3) {
                        case 2:
                            /* width_80p80 = true; */
                            /* fall through */
                        case 1:
                            ctx.width_160 = true;
                            break;
                        }
```

mariomaz replied:
Changed to support 80+80 MHz channels

arnout commented on common/beerocks/bwl/shared/nl80211_client_impl.cpp:
```diff
@@ -17,13 +21,276 @@
+ * @param freq frequency value in MHz.
+ * @return channel number.
+ */
+static uint16_t ieee80211_frequency_to_channel(uint32_t freq)
```
I think it's better to move beerocks::utils::wifi_freq_to_channel() to wireless_utils and use that.

mariomaz replied:
There is a FIXME in `beerocks::utils::wifi_freq_to_channel()` that I think has been fixed in `ieee80211_frequency_to_channel` borrowed from `iw`. @arnout / @tomereli  Can you please confirm which of two versions should I move to wireless_utils?

arnout replied:
The version you borrowed from `iw` is obviously better, but I still think there should be only one. So two commits are needed: one to move the function to wireless_utils, and a second one to fix it to correspond to how iw does it.

mariomaz replied:
Done

arnout commented on common/beerocks/bwl/shared/nl80211_client_impl.cpp:
```diff
@@ -17,13 +21,276 @@
+/**
+ * @brief Obtains the channel number that corresponds to given frequency value.
+ *
+ * This function has been borrowed from iw/util.c
```
When copying from somewhere else, please also copy the license blurb and copyright.

mariomaz replied:
Done

arnout commented on common/beerocks/bwl/shared/nl80211_client_impl.cpp:
```diff
@@ -17,13 +21,276 @@
 
+bool nl80211_client_impl::get_radio_info(const std::string &interface_name, sRadioInfo &radio_info)
+{
+    static struct nla_policy stats_policy[NL80211_STA_INFO_MAX + 1];
```
This isn't used. Cut&paste from update_stations_stats? :-)

mariomaz replied:
Removed

arnout commented on common/beerocks/bwl/shared/nl80211_client_impl.cpp:
```diff
@@ -17,13 +21,276 @@
+    stats_policy[NL80211_STA_INFO_TX_BITRATE]    = {NLA_NESTED, 0, 0};
+    stats_policy[NL80211_STA_INFO_RX_BITRATE]    = {NLA_NESTED, 0, 0};
+
+    static struct nla_policy rate_policy[NL80211_RATE_INFO_MAX + 1];
```
Also unused

mariomaz replied:
Removed

arnout commented on common/beerocks/bwl/shared/nl80211_client_impl.cpp:
```diff
@@ -17,13 +21,276 @@
+        bool width_40;
+        bool width_80;
+        bool width_160;
+    } ctx{};
```
(nitpick) why make this a struct? We can just as well use plain local variables in the closure.

mariomaz replied:
Long story. Code borrowed from `iw` creates a variable of type channels_ctx and passes it around because response can be split into several messages

arnout commented on common/beerocks/bwl/shared/nl80211_client_impl.cpp:
```diff
@@ -17,13 +21,276 @@
+                        LOG(DEBUG) << "NL80211_BAND_ATTR_HT_CAPA attribute is missing";
+                    }
+
+                    if (tb_band[NL80211_BAND_ATTR_HT_MCS_SET]) {
```
(nitpick) this is only useful/valid if HT_CAPA was set, and it's an ERROR if HT_CAPA was set and HT_MCS is not set.

mariomaz replied:
Changed to use NL80211_BAND_ATTR_HT_MCS_SET only if NL80211_BAND_ATTR_HT_CAPA was set and ignored otherwise

arnout commented on common/beerocks/bwl/shared/nl80211_client_impl.cpp:
```diff
@@ -17,13 +21,276 @@
+                        LOG(DEBUG) << "NL80211_BAND_ATTR_VHT_MCS_SET attribute is missing";
+                    }
+
+                    if (tb_band[NL80211_BAND_ATTR_FREQS]) {
```
I think we can treat it as an error (and return) if this is missing. That saves us some indentation depth below.

mariomaz replied:
Done

arnout commented on common/beerocks/bwl/shared/nl80211_client_impl.cpp:
```diff
@@ -17,13 +21,276 @@
+                                continue;
+                            }
+
+                            if (tb_band[NL80211_FREQUENCY_ATTR_FREQ]) {
```
```suggestion
                            if (tb_freq[NL80211_FREQUENCY_ATTR_FREQ]) {
```

However, the previous condition just checked that, so this is redundant.

mariomaz replied:
Fixed

arnout commented on common/beerocks/bwl/shared/nl80211_client_impl.cpp:
```diff
@@ -17,13 +21,276 @@
+                                tb_freq[NL80211_FREQUENCY_ATTR_DISABLED] ||
+                                (tb_freq[NL80211_FREQUENCY_ATTR_NO_IR] &&
+                                 tb_freq[__NL80211_FREQUENCY_ATTR_NO_IBSS]) ||
+                                tb_freq[__NL80211_FREQUENCY_ATTR_NO_IBSS]) {
```
`(a && b) || b` is equivalent to just `b`. In other words:
```suggestion
                                tb_freq[__NL80211_FREQUENCY_ATTR_NO_IBSS]) {
```

But maybe I missed something.

mariomaz replied:
Fixed

arnout commented on common/beerocks/bwl/shared/nl80211_client_impl.cpp:
```diff
@@ -17,13 +21,276 @@
+                            if (tb_freq[NL80211_FREQUENCY_ATTR_MAX_TX_POWER]) {
+                                channel.tx_pow_dBm =
+                                    0.01 *
+                                    nla_get_u32(tb_freq[NL80211_FREQUENCY_ATTR_MAX_TX_POWER]);
```
(nitpick) This way you convert to float and then back to int, so in the end you could just as well do
```suggestion
                                    nla_get_u32(tb_freq[NL80211_FREQUENCY_ATTR_MAX_TX_POWER]) / 100;
```

mariomaz replied:
Done

arnout commented on common/beerocks/bwl/shared/nl80211_client_impl.cpp:
```diff
@@ -17,13 +21,276 @@
+
+                            if (!tb_freq[NL80211_FREQUENCY_ATTR_NO_20MHZ]) {
+                                channel.bandwidth_MHz = 20;
+                            }
```
This will only be set for channels that are limited to 5 or 10MHz, which is probably not something we will ever want to use. So I think we can safely skip a channel that has NO_20MHZ set.

If not, we'll have to initialize bandwidth_MHz with a non-zero value, or it will *sometimes* be ignored...

mariomaz replied:
Done

arnout commented on common/beerocks/bwl/shared/nl80211_client_impl.cpp:
```diff
@@ -17,13 +21,276 @@
+                            }
+                            if (ctx.width_160 && !tb_freq[NL80211_FREQUENCY_ATTR_NO_160MHZ]) {
+                                channel.bandwidth_MHz = 160;
+                            }
```
I'm not sure how this will be used, but I believe we in the end will need a sWifiChannel for each separate bandwidth at which we can operate. At least, that's the assumption when constructing the EasyMesh messages (AP Radio Basic Capabilities). @kantera800 @YuriyMasechko @vitalybu please comment.

In practice, of course, the maximum bandwidth is sufficient since we can (probably) always operate at a lower bandwidth.

mariomaz replied:
Supported channels in this new implementation are obtained the same way as in previous implementation (which I guess will soon be replaced by this one).
I've also seen that nl80211 command returns a nested array of operating bands in `NL80211_ATTR_WIPHY_BANDS`. However, previous implementation considers just one band, the first one, and ignores the rest

mariomaz replied:
Modified as it was agreed during meeting: each channel includes a vector of supported_bandwidths.
I've also added a method to get maximum bandwidth supported

arnout reviewed (approved):
As often, some major comments but I don't think there's anything that can't be easily fixed.

Also as usual for @mariomaz: great documentation!

tomereli commented on common/beerocks/bwl/include/bwl/nl80211_client.h:
```diff
@@ -33,6 +51,116 @@ namespace bwl {
+         * Maximum channel bandwidth in MHz.
+         * Possible values are 0 (unknown), 20, 40, 80 and 160 MHz.
+         */
+        uint16_t bandwidth_MHz = 0;
```
Please don't mix camel case and lower snake case. The comment is enough actually so its fine to just call it `bandwidth`
```suggestion
        uint16_t bandwidth = 0;
```

mariomaz replied:
Ok

tomereli commented on common/beerocks/bwl/include/bwl/nl80211_client.h:
```diff
@@ -33,6 +51,116 @@ namespace bwl {
+         * Maximum transmission power in dBm.
+         * Obtained as 0.01 * NL80211_FREQUENCY_ATTR_MAX_TX_POWER
+         */
+        uint32_t tx_pow_dBm = 0;
```
```suggestion
        uint32_t tx_power = 0;
```

mariomaz replied:
Ok

tomereli commented on common/beerocks/bwl/include/bwl/nl80211_client.h:
```diff
@@ -33,6 +51,116 @@ namespace bwl {
+     * Information obtained with NL80211_CMD_GET_WIPHY command through a NL80211 socket.
+     * See <linux/nl80211.h> for a description of each field.
+     */
+    struct sRadioInfo {
```
There is already a radio info struct in `base_wlan_hal_types.h` - can we extend it and use it here instead?

mariomaz replied:
There is a radio info struct in `base_wlan_hal_types.h` and another one in `beerocks_message_structs.h`. @arnout / @tomereli  Which one should I use?
The same happens with the WiFi channel struct.

mariomaz replied:
Done as agreed during the meeting: use new sRadioInfo structure

tomereli reviewed (approved):
small nitpicks, looks good

mariomaz committed:
bcl: add 80+80 value to bandwidth enum

Add new value to bandwidth enum so 80+80 MHz channels can be
distinguished from 160 MHz channels.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
treewide: move beerocks::utils::wifi_freq_to_channel to son_wireless_utils

Move wifi_freq_to_channel, channel_to_freq and
channel_to_vht_center_freq from beerocks::utils to son_wireless_utils.

wifi_freq_to_channel and wireless_utils::which_freq could probably be
merged somehow, but this is left for another PR.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
bcl: fix wireless_utils::freq_to_channel with code from `iw`

Fix code in wireless_utils::freq_to_channel with code borrowed from
the `iw` command line tool.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

arnout commented on common/beerocks/bwl/include/bwl/nl80211_client.h:
```diff
@@ -33,6 +55,138 @@ namespace bwl {
+     * Information obtained with NL80211_CMD_GET_WIPHY command through a NL80211 socket.
+     * See NL80211_FREQUENCY_ATTR_* in <linux/nl80211.h> for a description of each field.
+     */
+    struct sChannelInfo {
```
I realize I'm royally late with giving this feedback, but according to our CODING_STYLE.md, we use lower snake case everywhere (except for cases of historical accident, like sMacAddress). So this should be `struct channel_info`.

(nitpick)

mariomaz replied:
Done

rmelotte commented on common/beerocks/bwl/dummy/nl80211_client_dummy.h:
```diff
@@ -31,6 +31,18 @@ class nl80211_client_dummy : public nl80211_client {
+     *
+     * @return True on success and false otherwise.
+     */
+    virtual bool get_radio_info(const std::string &interface_name, sRadioInfo &radio_info) override;
```
No need for virtual when there is already override (nitpick)

mariomaz replied:
Using `virtual` is an old habit from the times before C++11, when there was no `override`. It wasn't necessary either, but it was considered good practice to advertise programmers that a method was `virtual` without having to check the base class definition. I've also heard that some compilers might complain if `virtual` is not used in derived class, but that's not our case. 

Since `virtual` is being used together with `override` all over the code and it doesn't hurt, I'm going to leave it here as it is now. If we ever vote it in #democracy, I'll say do not use `virtual` if there is `override`. But I'll also say the use of `override` should be mandatory to avoid pitfalls.

rmelotte commented on common/beerocks/bwl/shared/nl80211_client_impl.cpp:
```diff
@@ -16,14 +41,236 @@
+        },
+        [&](struct nl_msg *msg) -> bool {
+            struct nlattr *tb[NL80211_ATTR_MAX + 1];
+            struct genlmsghdr *gnlh = (struct genlmsghdr *)nlmsg_data(nlmsg_hdr(msg));
```
No C-style cast, please use a static_cast if you still find the time/motivation to change it

mariomaz replied:
Done

rmelotte commented on common/beerocks/bwl/shared/nl80211_client_impl.h:
```diff
@@ -42,6 +42,18 @@ class nl80211_client_impl : public nl80211_client {
+     *
+     * @return True on success and false otherwise.
+     */
+    virtual bool get_radio_info(const std::string &interface_name, sRadioInfo &radio_info) override;
```
Same comment, no need for virtual here (there are a lot of places where it's done like this, but also others where we drop the override)

arnout replied:
I think we didn't come to a conclusion about it in the cppcheck PR. Vote in #democracy?
